src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +250,136 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	if (strstr(line, OUTPUT_BEGIN) == line)
+	{
+		/* If there is already a replay buffer open, it indicates a partial
+		 * transaction. Close it and start a new one.
+		 */
+		if (replayCtx->txnReplayBuffer != NULL)
+		{
+			fclose(replayCtx->txnReplayBuffer);
+		}
+
+		LogicalMessageMetadata metadata = { 0 };
+		if (!parseSQLAction((char *) line, &metadata))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		/* Create a txid as a symlink to the buffer file */
+		char txidFile[MAXPGPATH];
+		sformat(txidFile, MAXPGPATH, "%s/%d.sql", context->paths.dir, metadata.xid);
+
+		/* Open replaybuffer in w+ mode to truncate if the file already exists */
+		replayCtx->txnReplayBuffer = fopen_with_umask(txidFile,
+													  "w+", O_RDWR | O_TRUNC | O_CREAT,
+													  0644);
+		if (replayCtx->txnReplayBuffer == NULL)
+		{
+			log_error("Failed to open transaction buffer file \"%s\": %m",
+					  txidFile);
+			return false;
+		}
+
+		if (!create_symbolic_link(txidFile, context->paths.latesttxnfile))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		fformat(replayCtx->txnReplayBuffer, "%s\n", line);
+	}
+	else if (strstr(line, OUTPUT_COMMIT) == line)
+	{
+		if (replayCtx->txnReplayBuffer == NULL)
+		{
+			log_debug("Ignore COMMIT message received without a preceding BEGIN");
+			return true;
+		}
+
+		fformat(replayCtx->txnReplayBuffer, "%s\n", line);
+
+		/* Close replaybuffer to mark it as complete */
+		fclose(replayCtx->txnReplayBuffer);
+
+		/* It is time to replay the buffered transaction */
+
+		/* Open replaybuffer in read-only mode to use it as input stream */
+		FILE *txnReplayBufferForRead = fopen_read_only(context->paths.latesttxnfile);
+
+		if (txnReplayBufferForRead == NULL)
+		{
+			log_error(
+				"Failed to open transaction buffer file \"%s\" in readonly mode: %m",
+				context->paths.latesttxnfile);
+			return false;
+		}
+
+		ReadFromStreamContext readerContext = {
+			.callback = stream_replay_line,
+			.ctx = ctx
+		};
+
+		if (!read_from_stream(txnReplayBufferForRead, &readerContext))
+		{
+			log_error("Failed to read from replay buffer, "
+					  "see above for details");
+			return false;
+		}
+
+		fclose(txnReplayBufferForRead);
+
+		/* Remove the symlink to the replay buffer */
+		unlink_file(context->paths.latesttxnfile);
+
+		/* Early exit if we reached the end position */
+		*stop = context->reachedEndPos;
+
+		replayCtx->txnReplayBuffer = NULL;
+	}
+	else
+	{
+		if (replayCtx->txnReplayBuffer != NULL)
+		{
+			/*
+			 * We are in a transaction block, buffer all the messages
+			 * including KEEPALIVE, SWITCHWAL and ENPOS.
+			 */
+			fformat(replayCtx->txnReplayBuffer, "%s\n", line);
```

Comment:
Do we need to track `bool isTransaction` separately from the `FILE *` pointer being NULL, for ease of code maintenance and possibly advanced/intricate debugging?